### ISSUE TITLE
Can we do hyperparameter optimization while doing NAS?

### ISSUE THREAD

**Issue by curiousjit** [microsoft/nni#2591]
Short summary about the issue/question:

The built in one shot  NAS tuners are good for finding the best architecture, however the hyperparameters (#layers, #epochs or LR) are manually assigned. Is it possible to do NAS and hyperparameter optimization at the same time? 

nni Environment:

    nni version: 1.6
    nni mode(local|pai|remote): local
    OS: Linux
    python version: 3.6.10
    is conda or virtualenv used?: Conda
    is running in docker?: Yes

need to update document(yes/no): no

Anything else we need to know: no

**Comment by QuanluZhang** [comment]
@curiousjit this is a good feature request. We are working on this. One workaround for now is combining one-shot NAS with hyper-parameter tuning. You can start an NNI experiment, each trial is a one-shot NAS. In the search space of the experiment, you can configure the hyper-parameters that you want to try.

**Comment by curiousjit** [comment]
I am not sure how will that work @QuanluZhang . So if I want to do for DARTS, in config.yml:
`experimentName: nas_hpo_darts 
...
tuner:
...
trial:
    command: python3 search.py `
where search.py is from DARTS example.

And in search space file I can define the hyperparameters:
`"lr":{"_type":"choice", "_value":[0.1, 0.01, 0.001, 0.0001]},"layers":{"_type":"choice", "_value":[2,4,6]}`

If I start this NNI experiment, I am not sure how each trial can set the hyperparameter values in my one shot NAS (DARTS) code since they are already initialized in [search.py.](https://github.com/microsoft/nni/blob/master/examples/nas/darts/search.py). I will really appreciate your time if you can help me with this.

**Comment by QuanluZhang** [comment]
@curiousjit you can see DARTS's doc [here](https://nni.readthedocs.io/en/latest/NAS/DARTS.html#nni.nas.pytorch.darts.DartsTrainer), the trainer exposes some input arguments, for example `arc_learning_rate`. Then in each trial you give it a different value, this is like hyper-parameter tuning. In search space file, you give `arc_learning_rate` a range.

if the hyper-parameter that you want to tune is not exposed by the trainer, you can customize your DARTS trainer and expose the hyper-parameters through input argument. Then use NNI hyper-parameter tuning to tune one-shot NAS.

**Comment by curiousjit** [comment]
Thank you @QuanluZhang for your detailed reply. I can understand what you are trying to say but I think I have to do some changes. For example after looking at this [cifar10_pytorch](https://github.com/microsoft/nni/blob/master/examples/trials/cifar10_pytorch/main.py) hyper-parameter tuning example, I think I have to use [nni.get_next_parameter](https://github.com/microsoft/nni/blob/92b5fa145da1bef6576b2cc2ca327477f566c4c4/examples/trials/cifar10_pytorch/main.py#L189) in DARTS example code [search.py](https://github.com/microsoft/nni/blob/master/examples/nas/darts/search.py) to input a parameter (arc_learning_rate) from my search space file to DARTS trainer.

Currently I have attached my search_space.json and config.yml and if you think the above idea is correct, kindly let me know and I will work on it. Thank you again for your time.
[attachments.zip](https://github.com/microsoft/nni/files/4843764/attachments.zip)

**Comment by QuanluZhang** [comment]
@curiousjit  yes, your solution is correct. feel free to feedback if you have further problems.

**Comment by curiousjit** [comment]
Thank you @QuanluZhang for your response. I will keep this issue open just in case if I have any additional questions while implementing it. As soon as I do it, I will close the issue too.

**Comment by QuanluZhang** [comment]
@curiousjit great. welcome to share your experience and feedbacks with us.

**Comment by roya0045** [comment]
Is this a thing that is viable? From my understanding as we change the structure of a model we also change slightly the impact on some parameters. Once we are closer to an optimal model it may be more viable to search for optimal parameters for a given model and other similar structures. And won't tuning some parameter also bias the search to find a 'local optimum', IE an effficient small model rather than an overall better model in the long term?

I have not done such trials but I think it might be something to try. Let me know if I'm wrong in this case. Thanks!

**Comment by QuanluZhang** [comment]
@roya0045 , I agree with your statement "we change the structure of a model we also change slightly the impact on some parameters". Both architecture and hyper-parameters have impact on final accuracy. Sometimes architecture has more impact while sometimes hyper-parameters have more impact. Different architectures may need different hyper-parameters to achieve their highest accuracy, thus, it is reasonable to jointly optimize architecture and hyperparameter. It is now a hot research topic.

**Comment by scarlett2018** [comment]
@QuanluZhang - with the release of Retiarii, I suppose this ability is enabled. may we close this issue?

**Comment by QuanluZhang** [comment]
> @QuanluZhang - with the release of Retiarii, I suppose this ability is enabled. may we close this issue?

yes, this ability will be supported in v2.3

**Comment by QuanluZhang** [comment]
@curiousjit , co-optimizing hyper-parameter and neural architecture is achieved by `nn.LayerChoice` and `nn.ValueChoice`. Jointly optimizing hyper-parameters in training logic will be supported in v2.4

**Comment by curiousjit** [comment]
> @curiousjit , co-optimizing hyper-parameter and neural architecture is achieved by `nn.LayerChoice` and `nn.ValueChoice`. Jointly optimizing hyper-parameters in training logic will be supported in v2.4

I see. Thank you for the info! Looking forward to it.
